lxd/storage_volumes.go

```diff
@@ -2727,3 +2727,84 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
 	revert.Success()
 	return operations.OperationResponse(op)
 }
+
+// getRemoteVolumeNodeInfo figures out the cluster member on which the volume with the given name is defined. If it is
+// the local cluster member it returns nil and no error. If it is another cluster member it returns a db.NodeInfo containing
+// the name and address of the remote member. If there is more than one cluster member with a matching volume name, an
+// error is returned.
+func getRemoteVolumeNodeInfo(ctx context.Context, s *state.State, poolName string, projectName string, volumeName string, volumeType int) (*db.NodeInfo, error) {
+	localNodeID := s.DB.Cluster.GetNodeID()
+	var err error
+	var nodes []db.NodeInfo
+	var poolID int64
+	err = s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+		poolID, err = tx.GetStoragePoolID(ctx, poolName)
+		if err != nil {
+			return err
+		}
+
+		nodes, err = tx.GetStorageVolumeNodes(ctx, poolID, projectName, volumeName, volumeType)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil && err != db.ErrNoClusterMember {
+		return nil, err
+	}
+
+	// If volume uses a remote storage driver and so has no explicit cluster member, then we need to check
+	// whether it is exclusively attached to remote instance, and if so then we need to forward the request to
+	// the node whereit is currently used. This avoids conflicting with another member when using it locally.
+	if err == db.ErrNoClusterMember {
+		// GetStoragePoolVolume returns a volume with an empty Location field for remote drivers.
+		var dbVolume *db.StorageVolume
+		err = s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
```

Comment:
I believe this transaction could be avoided by moving the conditional  tx.GetStoragePoolVolume call into the earlier transaction based on the error value returned by the earlier DB calls.

Reply:
Fixed.